### PR TITLE
Enable Ansible control node management for registered systems

### PR DESCRIPTION
**Pull request by meaksh**
Read it [here](https://github.com/uyuni-project/uyuni-rfc/blob/master-enable-ansible-control-node-in-registered-system/accepted/00000-enable-ansible-control-node-in-registered-system.md)

### CODE REVIEW

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:74]: I think this case can be assimilated to having a SUSE Manager Proxy, where the clients can only reach the Server through the Proxy but not directly. Deploying the SaltSSH seems OK to me.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:69]: I don't necessarily see this is as a problem, at least initially

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:76]: Is this something that's usually defined? Or is this something very special we would be requiring? That's for me the main factor to decide for one or the other alternative.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:78]: We will be using salt-ssh but clients may be salt-minions or salt-ssh clients, right?

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:94]: Should we also provide an easy way to trigger this from the UI? E. g. by adding Ansible playbooks directly to configuration channels (without the Salt state in the middle, we can autogenerate that state)?

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:120]: Say I have 1,000 Ansible clients and I want to bootstrap them with salt-ssh. Will it take a lot more time than if I bootstrap them with the normal bootstrap? To me, it's the same, the only additional step is getting the SSH key from Ansible. Or am I missing something?

**Review by paususe — COMMENTED** [review]
Left a few comments and questions

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (mcalmer) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:94]: Maybe not in the first step. The problem is, that playbooks have the targets where it gets applied included. This cannot be changed without editing playbooks which is not planned yet. This topic along would need an own RFC from my point of view.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (agraul) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:78]: That's my understanding, yes. Just like regular boostrapping which is executed over salt-ssh for both salt-ssh and salt-minion clients

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:94]: > Maybe not in the first step

OK

> The problem is, that playbooks have the targets where it gets applied included. This cannot be changed without editing playbooks which is not planned yet. This topic along would need an own RFC from my point of view.

OK. Maybe an opportunity to move the editor to VS Codium or alike :-)

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (mcalmer) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:120]: Using our UI you can only register 1 and even with API you register one after another.

Here I would expect some kind of parallel onboarding and this requires batches and other things.
Performance is a possible problem. It depends on the implementation.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (meaksh) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:76]: This NOTE actually applies for both approaches. The SSH keys or credentials to connect with the Ansible managed clients should be defined in the Ansible inventory.

If an user do not specify any user/pass/sshkey in the Ansible inventory to connect with different Ansible managed clients, then Ansible will use the default SSH identity of the user that executes `ansible` CLI to try to connect to the Ansible managed clients.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (meaksh) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:78]: Yes, that's correct.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (meaksh) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:120]: It should take the same time in both cases. In the end, the "Boostrapping UI" page is basically also doing "salt-ssh" execution under the hood.

As you mentioned, the only additional step would be to fetch the SSH keys from the Ansible control node first, and not necessary for each client to bootstrap but just once at the time of getting the Ansible inventory.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (moio) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:102]: It is my understanding that we do not intend to store the list of playbooks in Postgres at least for now. The list is retrieved live at page load, displayed as a list with a button to run each, and once the button is click the action is scheduled.

Correct?

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (moio) [inline accepted/00000-enable-ansible-control-node-in-registered-system.md:128]: I'd recommend trying to put down pros and cons to the extent possible. At least I would recommend making it clear that the basic premise of this design (controller is registered) could potentially be limiting compared to the other design.

**Review by moio — COMMENTED** [review]
@meaksh thanks for the quick write up!

My main remark with the current text comes from the following considerations:
 - we do have mass registration as an open general problem since a while, [in Public Cloud contexts](https://github.com/SUSE/spacewalk/issues/10884) and otherwise. [We kind of suggest working around it in the documentation](https://github.com/uyuni-project/uyuni-docs/blob/master/modules/large-deployments/pages/operation-reqs.adoc#salt-client-onboarding-rate). This could be the right occasion to solve it once and for all, by providing one good mechanism either leveraging the existing 1-system-at-a-time API or introducing a new one, and solve a general day-1 problem many users face. Fact that systems to onboard come from an Ansible controller rather than someplace else does not change the problem's essence - I would not want to have different mechanisms for different instances of the same problem
 - onboarding Ansible end clients is not necessarily desirable in the first place. For starters, users might only really want to just running playbooks. Second, clients might be on platforms we do not support. Third, onboarding can always happen via running a playbook - simply to run a bootstrap script! Fourth, users might not want to have salt-minion co-installed for non-technical reason. Fifth and most important to me, they might not want to adapt their playbooks to make that happen. In fact, when a decision is made to have a system managed both with Salt and Ansible, extra care in writing both playbooks and states, otherwise they will conflict, [particularly in the area of repo management as our own internal sysadmins experimented a while ago](http://mailman.suse.de/mlarch/SuSE/suse-manager/2018/suse-manager.2018.10/msg00015.html)
- we are a bit short on time and want to deliver a minimal increment fast

Keeping those factors in mind, I recommend moving the whole onboarding problem to a separate RFC, to be implemented once we added the ability to run playbooks, which is the core value add at the moment. Text you have created here should be kept (as it's valuable!) in a "next steps" section.

**Comment by paususe** [comment]
> @meaksh thanks for the quick write up!
> 
> My main remark with the current text comes from the following considerations:
> 
>     * we do have mass registration as an open general problem since a while, [in Public Cloud contexts](https://github.com/SUSE/spacewalk/issues/10884) and otherwise. [We kind of suggest working around it in the documentation](https://github.com/uyuni-project/uyuni-docs/blob/master/modules/large-deployments/pages/operation-reqs.adoc#salt-client-onboarding-rate). This could be the right occasion to solve it once and for all, by providing one good mechanism either leveraging the existing 1-system-at-a-time API or introducing a new one, and solve a general day-1 problem many users face. Fact that systems to onboard come from an Ansible controller rather than someplace else does not change the problem's essence - I would not want to have different mechanisms for different instances of the same problem

Mass registration is a nice enhancement but not a must have for the MVP

> 
>     * onboarding Ansible end clients is not necessarily desirable in the first place. For starters, users might only really want to just running playbooks. 

We are not interested in this case. Uyuni is a systems management solution. Users who do not want to manage do not need Uyuni, they can stay with Ansible.

> Second, clients might be on platforms we do not support. 

That's a can of worms. We are not interested in that case. If Ansible controller X has 1,000 clients in inventory and 300 of those clients are on platforms we do not support, then only the other 700 clients should be registered with Uyuni.

> Third, onboarding can always happen via running a playbook - simply to run a bootstrap script! 

OK

> Fourth, users might not want to have salt-minion co-installed for non-technical reason. 

This is an opinionated Ansible integration. If they do not watn salt-minion, then they can use salt-ssh. If they do not want salt-minion nor salt-ssh, then they should not use Uyuni but stay with Ansible.

> Fifth and most important to me, they might not want to adapt their playbooks to make that happen. In fact, when a decision is made to have a system managed both with Salt and Ansible, extra care in writing both playbooks and states, otherwise they will conflict, [particularly in the area of repo management as our own internal sysadmins experimented a while ago](http://mailman.suse.de/mlarch/SuSE/suse-manager/2018/suse-manager.2018.10/msg00015.html)

They will need to take care of that. Conflicts and uncertainly can happen today too, e. g. if you subscribe to configuration channels with directly conflicting orders.

Again, this is an opinionated Ansible integration. Trying to support every possible case for everybody will be a nightmare to support and will lead to surprises, e. g. 
"User: Why is this system not being updated"
"Uyuni developers: Because you are not managing it, just running playbooks on it"
"User: oh, right, I had forgot I have a mix of managend and non-managed systems"

I do not want that to happen.


> 
>     * we are a bit short on time and want to deliver a minimal increment fast
> 
> 
> Keeping those factors in mind, I recommend moving the whole onboarding problem to a separate RFC, to be implemented once we added the ability to run playbooks, which is the core value add at the moment. Text you have created here should be kept (as it's valuable!) in a "next steps" section.

OK

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (paususe) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:128]: Why would it be limiting? On the contrary, I see this proposal providing a much better integration than the other one. The cases the registered Ansible controller makes more difficult are cases are do not really want to get into since they would provide a worse user experience or a more difficult to support integration.

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (meaksh) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md:102]: That's correct, at least for now.

Beside of listing the playbooks, we probably also want the user can display the content of the playbook (as readonly) before triggering it clicking the button, but this can be done in live without storing any content in the DB:

```console
# salt ANSIBLE_CONTROL_NODE cp.get_file_str /path/to/the/playbook
```

File: accepted/00000-enable-ansible-control-node-in-registered-system.md
Comment (moio) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md]: Perhaps that was a wrong choice of words - what I mean is "less general" - it just poses an additional requirement (registration) that the other approach does not have.

Comment (ncounter) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md]: I'd not define it a "worse or better" user experience though, it would be just different because different is the approach and the resulting number of things one can do with an onboarded/imported system.

First bunch of differences I see (@meaksh up to you to organize them in pros and cons) :
- A registered system can be managed, an imported one can just show up in the list
- A registered system needs to be of a supported flavor by us, an imported one can be of any kind
- Replacing a system would means deleting the profile and re-registering it, an imported one simply need to keep the same IP and can be swapped

On the other side I see the "limiting" word as another way of saying "stricter integration", which is fine as long as it is intended to be. I'd say the other RFC was born on a different perspective, now we explored the two paths deeply, and we see differences. Probably it was worth for the awareness sake and for the future. We can always add the "pre-step" of import systems only. I agree it would be good to keep it as a note in the current RFC here to have a full picture in one view.

Comment (meaksh) [reply accepted/00000-enable-ansible-control-node-in-registered-system.md]: Thanks for the feedback here!

I've just added some notes in the RFC text to highlight some of the relevant different between the two approaches without really evaluating them in Pros/Cons. Just for the record, so we can even take it into consideration for future enhacements of the current integration proposed in the present RFC.

**Comment by paususe** [comment]
@meaksh is there any reason not to merge this RFC already?

**Comment by meaksh** [comment]
@paususe I was waiting for a formal approval here. Ping to PR reviewers to get the approval before merging. 

@moio @mcalmer @ncounter @agraul

**Review by mcalmer — APPROVED** [review]

**Review by agraul — APPROVED** [review]

**Review by ncounter — APPROVED** [review]
:+1:

**Comment by meaksh** [comment]
I'm merging this now. If there are more consideration we can tackle them in followup PRs.

Thanks everybody for participating here and for all the feedback you provided!

**Pull request merged**
